Refactor #6489 JumpToFirstUnplayedItem to SelectFirstUnwatchedItem

Carcharius opened the pull request:
As per @xhaggi request in comments for #6489 the first commit changes the settings and method names from videolibrary.jumptofirstunplayeditem and JumpToFirstUnplayedItem to videolibrary.selectfirstunplayeditem and SelectFirstUnplayedItem.

The second commit changes the setting from a boolean to an integer in order to allow future extension of the setting options to allow finer grained control of how the item should be selected e.g. whether to skip over the All Seasons or Specials items.

**UPDATE**
Following discussion with @xhaggi  and @mkortstiege and other comments I have amended the changes as follows:
- I've backed out the previous changes to CGUIMediaWindow and CGUIViewState (and derivatives).
- I've reimplemented as part of CGUIWindowVideoNav
- It is no longer applicable just when episodes are sorted by episode number. It now works by finding the lowest unwatched episode number (or season number if at season level), no matter what the sort type is.

MartijnKaijser commented on language/English/strings.po:
```diff
@@ -11579,7 +11579,17 @@ msgctxt "#21470"
+msgid "Yes"
+msgstr ""
+
+#empty strings from id 21473 to 21601
```
drop these and find the other YES/NO labels in strings

Carcharius replied:
Will do. Didn't realise it was ok to reuse existing strings.

Carcharius commented:
As per discussion for #6489 this PR shouldn't be merged just yet.

@xhaggi suggests that CGUIMediaWindow isn't the correct place to handle this, so I will look into using CGUIWindowVideoBase instead and update this PR when I am done.

Sorry if I'm making a mess :blush:

xhaggi commented:
@Carcharius sounds good, thanks

I've talked to @mkortstiege and we both think we should use "unwatched" instead of "unplayed" e.g. `selectfirstunwatcheditem`.

And could you explain what you mean with "allow finer grained control of how the item should be selected"? Do you have an example of what you intend to add later.

xhaggi commented:
After a second talk with @mkortstiege it would be great if you could add this kind of "default selection" in a more generic way. For example:

Default item selection:
* Default (which means the first item or the last selected item)
* First unwatched Item (if episodes the first episode, if movies the first unwatched movie)
* Last watched item (could be an option, or couldn't?)

Therefor the setting/methods etc. should be renamed to `defaultitemselection`.
Besides we could also add an option to configure how often we do this selection "Initially" or "Always" like suggested by @HitcherUK

Carcharius commented:
I can certainly look at making it work in the way suggested.

I used lastplayeditem because initially I thought it could also be applied to music at some point and last watched didn't seem suitable in that context.

My ideas on finer control were to, as mentioned, add options to control if All Seasons and/or Specials should be included.

After @HitcherUK comments I had started to think about having 2 settings. The first being the equivalent of what I have now but expanded to be Never, Initially or Always.

The second setting would be a child of the first, possibly only visible at higher levels. That would be something like "Include All Seasons and Specials", with options of Neither, Both, Just All Seasons, Just Specials.

Do you think I should continue to consider the sort type?

At the moment, at @mkortstiege suggestion, it only applies to episodes if the sort is by episode number.

If I am expanding to include movies, is there a specific sort type I should target? (Am away from Kodi atm so not sure what sort types apply to movies).

Would CGUIWindowVideoBase still be the correct place to look to apply the selection?

HitcherUK commented:
Not sure I can see the benefit for movies myself as they're not ordered in a way that you'd want to watch the next one in the list. Can you explain where you'd want to do this?

Carcharius commented:
@HitcherUK I agree with you. I mentioned movies because @xhaggi did.

I suppose it might make sense for movie sets maybe. You might have watched Alien and want it to automatically select Aliens for you.

xhaggi commented:
> Not sure I can see the benefit for movies myself as they're not ordered in a way that you'd want to watch the next one in the list. Can you explain where you'd want to do this?

If this makes no sense drop it. I thought about an option "last watched" which might be work for movies.
I'm fine if this is only applied to tvshows view. @Carcharius if so please name the setting "tvshowsdefaultselection".

> Do you think I should continue to consider the sort type?

If it's possible to select the episode based on its episode number (last watched episode was 5, next episode should be 6), I would say no.

> Would CGUIWindowVideoBase still be the correct place to look to apply the selection?

yes

xhaggi commented:
> My ideas on finer control were to, as mentioned, add options to control if All Seasons and/or Specials should be included.
> 
> After @HitcherUK comments I had started to think about having 2 settings. The first being the equivalent of what I have now but expanded to be Never, Initially or Always.
> 
> The second setting would be a child of the first, possibly only visible at higher levels. That would be something like "Include All Seasons and Specials", with options of Neither, Both, Just All Seasons, Just Specials.

Only to understand this right you want something like:

"Select first unwatched TV show season/episode"
* Never/Off
* Initially
* Always

"Include All Seasons and Specials" < only available if you choose "Initially" or "Always"
* Neither
* Both
* Just All Seasons
* Just Specials.

xhaggi commented:
@Carcharius after a second look at CGUIWindowVideoBase things should go to CGUIWindowVideoNav. We don't need this for CGUIWindowVideoPlaylist which inherited from CGUIWindowVideoBase.

xhaggi commented on xbmc/windows/GUIMediaWindow.cpp:
```diff
@@ -860,7 +860,8 @@ bool CGUIMediaWindow::Update(const std::string &strDirectory, bool updateFilterP
 
     // Check if we should select the first unplayed item
-    if (m_guiState.get()->JumpToFirstUnplayedItem())
+    int iSelectFirstUnplayed = m_guiState.get()->SelectFirstUnplayedItem();
```
you can drop the `.get()` and use `m_guiState->SelectFirstUnplayedItem()`

xhaggi commented on xbmc/view/GUIViewState.cpp:
```diff
@@ -395,9 +395,9 @@ bool CGUIViewState::AutoPlayNextItem()
 }
 
-bool CGUIViewState::JumpToFirstUnplayedItem()
+int CGUIViewState::SelectFirstUnplayedItem()
```
drop this and the declaration after you moved the selection code to CGUIWindowVideoNav

mkortstiege commented:
@xhaggi, maybe i misread but how should marking the next episode work if we´re not sorting by episode or airdate?

Carcharius committed:
[video] refactor select first unwatched season/episode item

Undo previous changes made to add selection of first unplayed item.

Add settings videolibrary.tvshowselectfirstunwatcheditem (renamed from videolibrary.jumptofirstunplayeditem) and videolibrary.tvshowsincludeallseasonsandspecials

Add unwatched item selection to CGUIWindowVideoNav.

Carcharius commented:
I've updated the PR, please see the original description for explanation.

mkortstiege commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,114 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+{
+  if (m_vecItems->IsVideoDb())
+  {
+    SortBy sortBy = m_guiState->GetSortMethod().sortBy;
```
leftover?

Carcharius replied:
Yep - will nuke it and update

mkortstiege commented on system/settings/settings.xml:
```diff
@@ -345,10 +345,33 @@
+            <options>
+              <option label="20420">0</option> <!-- Never -->
+              <option label="21471">1</option> <!-- Initially -->
+			  <option label="20422">2</option> <!-- Always -->
```
indentation

mkortstiege commented on system/settings/settings.xml:
```diff
@@ -345,10 +345,33 @@
+            <options>
+              <option label="21474">0</option> <!-- Neither -->
+              <option label="21475">1</option> <!-- Both -->
+			  <option label="21476">2</option> <!-- Just All Seasons -->
```
indentation ;)

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+{
+  if (m_vecItems->IsVideoDb())
+  {
+    NODE_TYPE NodeType = CVideoDatabaseDirectory::GetDirectoryChildType(m_vecItems->GetPath());
```
variable should starts with lower case.

Carcharius replied:
hmm that was a copy and paste from existing code in GUIViewStateVideo.cpp, line 118

xhaggi replied:
okay, but we should change it according to our code format ;)

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+int CGUIWindowVideoNav::GetFirstUnwatchedItemIndex(bool includeAllSeasons, bool includeSpecials)
+{
+  int iIndex = 0;
+
```
minor, but please remove this empty line.

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+  int iUnwatchedSeason = INT_MAX;
+
+  // Run through the list of items and find the season number of the first season with unwatched episodes
+  for (int i = 0; i < m_vecItems->Size(); ++i)
```
`m_vecItems` use a zero-based index, we shouldn't increment it before, so this should be `i++`

tamland replied:
that's.. not how loops works. should always prefer ++i

xhaggi replied:
you are right don't know why I thought it works like I said ;)

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+    }
+  }
+
+  NODE_TYPE NodeType = CVideoDatabaseDirectory::GetDirectoryChildType(m_vecItems->GetPath());
```
same here, should starts with lower case

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
   }
 }
 
+int CGUIWindowVideoNav::SelectFirstUnwatchedItem()
```
should we use an enum instead on an int here?

```
enum SelectFirstUnwatchedItem
{
  NEVER = 0,
  INITIALLY = 1,
  ALWAYS = 2
}
```

Carcharius replied:
Where should the enum be declared? Is it ok to declare it locally to CGUIWindowVideoNav?

xhaggi replied:
sure we only need it there for now

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+  return 0;
+}
+
+int CGUIWindowVideoNav::IncludeAllSeasonsAndSpecials()
```
enum?

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+
+      // Is the episode unwatched, and is its epsiode number lower than the currently
+      // identified first unwatched episode
+      if (pTag->m_playCount == 0 && pTag->m_iEpisode < iUnwatchedEpisode)
```
please add the above check for the season to this conditions.

```
if (pTag->m_iSeason == iUnwatchedSeason && pTag->m_playCount == 0 && pTag->m_iEpisode < iUnwatchedEpisode)
```

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+
+    // Now run through the list of items and check episodes from the season identified above
+    // to find the first (lowest episode number) unwatched epsisode.
+    for (int i = 0; i < m_vecItems->Size(); ++i)
```
same here use `i++`

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.h:
```diff
@@ -55,7 +55,13 @@ class CGUIWindowVideoNav : public CGUIWindowVideoBase
+
+  virtual int SelectFirstUnwatchedItem();
+  virtual int IncludeAllSeasonsAndSpecials();
+  virtual int GetFirstUnwatchedItemIndex(bool includeAllSeasons, bool includeSpecials);
```
please change this methods to be private

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+  int iSelectFirstUnwatched = SelectFirstUnwatchedItem();
+  if (iSelectFirstUnwatched > 0)
+  {
+    int iItem = m_viewControl.GetSelectedItem();
```
if you change this var to be a bool and rename the var to `bIsItemSelected` you can drop the comment 2 lines below.

```
bool bIsItemSelected = (m_viewControl.GetSelectedItem() > 0);
```

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,112 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+  if (NodeType == NODE_TYPE::NODE_TYPE_EPISODES)
+  {
+    iIndex = 0;
+
```
same here drop the empty line

xhaggi commented:
@Carcharius apart from my comments this looks way better, thanks, and hope you can fix the rebase issue.

MartijnKaijser commented:
@jjd-uk can you check the new strings?

Carcharius commented:
@xhaggi or anyone else.

Could somebody please help me with the rebase issue?
I don't know what I did wrong there.

xhaggi commented:
sure, do the following:

```
git rebase -i 9b55f9b
// comment out the first 2 commits in your editor
// save and exit your editor

// not fetch the xbmc remote 
git fetch -p xbmc (or your name for the xbmc remote repo)

// rebase against the xbmc/master
git rebase xbmc/master (or your name for the xbmc remote repo master)

// force push
git push -f

```

xhaggi commented on system/settings/settings.xml:
```diff
@@ -345,10 +345,33 @@
           <control type="spinner" format="string" />
         </setting>
-        <setting id="videolibrary.jumptofirstunplayeditem" type="boolean" label="21416" help="21466">
+        <setting id="videolibrary.tvshowselectfirstunwatcheditem" type="integer" label="21416" help="21466">
```
did you forgot an "s" by tvshow(s)select... ;)

Carcharius replied:
I thought about putting it in by didn't really like the double "s" that resulted, looked like a typo ;)

Will put it back

xhaggi replied:
I don't like to differ between the other settings where we use "tvshows".

Carcharius commented:
Thanks for the guidance. 

I've (hopefully) resolved the rebase issue, addressed the comments, rebased on upstream/master and squashed my commits (just because I knew I was going to get asked to do so at some point :stuck_out_tongue: )

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,114 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
   }
 }
 
+SELECT_FIRST_UNWATCHED_ITEM CGUIWindowVideoNav::SelectFirstUnwatchedItem()
```
i would prefer camel case for enum names instead of all upper.

Carcharius replied:
I had it that way originally, but it conflicted with the name of the method, SelectFirstUnwatchedItem, and Visual Studio didn't like it when trying to cast the int received from CSettings to the enum.

xhaggi replied:
then we need to use a different name

xhaggi replied:
or BTW change the method name, it sound not like a getter for the setting option

Carcharius replied:
Have prefixed the enum names with e - so eSelectFirstUnwatchedItem - hope that is acceptable.

xhaggi replied:
no ;) .. first should be upper like a class name.

xhaggi replied:
should we prefix the methods with `GetSetting..`?

Carcharius replied:
You're just doing it to annoy me now aren't you? ;) lol

I did look to see if there were any existing enums prefixed with e and there are so I thought it would be ok.

But I have removed the prefix and changed the methods to GetSetting... as suggested

xhaggi replied:
sorry for that but most of the enums follows the same code convention as we have for classes and I never know that we have enums prefixed with "e" ;)

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,114 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+  return SELECT_FIRST_UNWATCHED_ITEM::NEVER;
+}
+
+INCLUDE_ALL_SEASONS_AND_SPECIALS CGUIWindowVideoNav::IncludeAllSeasonsAndSpecials()
```
same here

xhaggi commented on xbmc/video/windows/GUIWindowVideoNav.cpp:
```diff
@@ -268,6 +268,114 @@ std::string CGUIWindowVideoNav::GetQuickpathName(const std::string& strPath) con
+    if (nodeType == NODE_TYPE_SEASONS || nodeType == NODE_TYPE_EPISODES)
+    {
+      int iValue = CSettings::Get().GetInt("videolibrary.tvshowsselectfirstunwatcheditem");
+      if (iValue >= SELECT_FIRST_UNWATCHED_ITEM::NEVER && iValue <= SELECT_FIRST_UNWATCHED_ITEM::ALWAYS)
```
you don't need to prefix the enum name.

Carcharius replied:
If I don't then NEVER conflicts with another value in CGUIControl::GUISCROLLVALUE

xhaggi replied:
ahh sorry it's fine then

xhaggi commented:
> Thanks for the guidance.

np ;)

It would also be nice if you could rename your commit message to something more reasonable
`[video] refactor select first unwatched season/episode item`

Carcharius commented:
I've pushed the latest changes.
* renamed the enums and the settings getter methods
* reverted back to iterating m_vecItems with `++i` as per comments from @tamland and @xhaggi 
* used @xhaggi suggestion for the commit message.

xhaggi commented:
Thank you very much indeed ;)

jenkins build this please

xhaggi commented:
after @jjd-uk checked the strings it's good to merge.
@mkortstiege any objections?

mkortstiege commented:
Looks good. No objections.

jjd-uk commented on language/English/strings.po:
```diff
@@ -11324,7 +11324,7 @@ msgstr ""
 #: system/settings/settings.xml
 msgctxt "#21416"
-msgid "Jump to first unwatched TV show season/episode"
+msgid "Select first unwatched TV show season/episode"
```
Add **the** to give:

"Select **the** first unwatched TV show season/episode"

jjd-uk commented on language/English/strings.po:
```diff
@@ -11579,7 +11579,43 @@ msgctxt "#21470"
+#: system/settings/settings.xml
+msgctxt "#21471"
+msgid "Initially"
+msgstr ""
```
This for "Select the first unwatched TV show season/episode" - "Never/Initially/Always" right?

I don't understand what "Initially" does in this context, can you explain setting Initially does?

HitcherUK replied:
@jjd-uk It will select the first unwatched item 'initially' but if the user then focuses on another item then leaves and re-enters the focus will still be on their last selection. Using 'always' upon re-entry will select the first unwatched item again.

MartijnKaijser replied:
there should be descriptions added for translators (for all new strings preferably)

Carcharius replied:
I'll give it a try :wink: 

Say you have a tv show with a number of seasons and each season has a number of episodes. You go into season 2 and move along to episode 7. Then you go back to the season level for some reason but decide to go back into season 2.

As Kodi stands at the moment, when you go back in to season 2 episode 7 will still be selected.

With this PR in place.

If you have the setting set to Never, nothing will change the current functionality will remain.

If you have it set to Initially then, on that first entry into season 2, the first unwatched episode will be selected. Lets say that is episode 5. Then you decide to select episode 7, go back to seaon level and then back into season 2. In this case when you go back into season 2, because there is history recorded (that you had episode 7 selected) episode 7 will be selected again.

If you have the setting set to Always, then every time you enter the episode list the first unwatched episode will be selected, whether there is history recorded or not. So in the previous example, both times you enter season 2 the unwatched episode 5 would be selected.

I hope that makes sense!

In my original PR the "Initially" behaviour was the only behaviour. But @HitcherUK requested the facility for the first unwatched item to be selected every time.

Carcharius replied:
@HitcherUK explained it much better and more concisely than I did.

@MartijnKaijser where should the descriptions be placed? Are there existing examples so I can see how things should be?

MartijnKaijser commented:
See
https://github.com/xbmc/xbmc/pull/6505/files#diff-7f1c48946a74807ac73a78e6a5f8d13cR11551

Carcharius commented:
I've added the missing "the" and also added some descriptions to the strings.

@MartijnKaijser I hope that's what you were looking for.

jjd-uk commented:
Thanks for the explanations of what "Initially" does, makes sense now I think however not keen on using the word "Initially" for this though.

Would something like "First view only" be more descriptive?

Carcharius commented:
How about "On first entry"

jjd-uk commented:
Yep that would do it for me I think, but let's see if there's any differing opinions before you go ahead and change it.

mkortstiege commented:
"On first entry" sounds good to me and there's also the description that ensures the user knows that this means.

Carcharius commented:
I've updated and changed `Initially` to 'On first entry'. I've also expanded the description for the setting to try and explain how the options work.

jjd-uk commented:
Thanks, all is good on the strings side now :)

Carcharius commented:
Great. Once it is merged I'll update the wiki with the new details

xhaggi commented:
@Carcharius thanks for your work on this and thanks everyone else for reviewing/testing ;)

MartijnKaijser commented:
needs to be rebased on master

Carcharius commented:
Rebased and pushed :smile:

mkortstiege commented:
Thanks! jenkins build this please.

HitcherUK commented:
@Carcharius @xhaggi @mkortstiege @Montellese 

Just noticed that when you finish watching something it gets marked as watched but the next item no longer gets focused in the same way that it does if you simply mark it as watched. Could this have something to do with that?
